File: src/test/java/com/timeback/ui/viewmodel/ReportViewModelTest.java
package com.timeback.ui.viewmodel;

import com.timeback.ui.domain.model.ReportData;
import com.timeback.ui.domain.model.ReportPeriod;
import com.timeback.ui.domain.model.ScreenState;
import com.timeback.ui.fake.FakeFeatureGateway;
import com.timeback.ui.feature.report.ReportViewModel;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import static org.junit.jupiter.api.Assertions.*;

/**
 * ReportViewModel 단위 테스트.
 */
@ExtendWith(LiveDataTestExtension.class)
class ReportViewModelTest {

    private FakeFeatureGateway gateway;
    private ReportViewModel viewModel;

    @BeforeEach
    void setUp() {
        gateway = new FakeFeatureGateway();
    }

    @Test
    void loadReport_withBaselineReady_shouldReturnContent() throws Exception {
        gateway.baselineReady = true;
        viewModel = new ReportViewModel(gateway);
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        assertNotNull(state);
        assertInstanceOf(ScreenState.Content.class, state);

        Object data = ((ScreenState.Content) state).getData();
        assertInstanceOf(ReportData.class, data);
        ReportData rd = (ReportData) data;
        assertEquals(ReportPeriod.WEEKLY, rd.getPeriod());
        assertTrue(rd.isBaselineComparable());
        assertFalse(rd.getGoalSummaries().isEmpty());
    }

    @Test
    void loadReport_withoutBaseline_shouldReturnEmpty() throws Exception {
        gateway.baselineReady = false;
        viewModel = new ReportViewModel(gateway);
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        assertNotNull(state);
        assertInstanceOf(ScreenState.Empty.class, state);
    }

    @Test
    void changePeriod_shouldUpdateReport() throws Exception {
        gateway.baselineReady = true;
        viewModel = new ReportViewModel(gateway);
        Thread.sleep(300);

        viewModel.changePeriod(ReportPeriod.DAILY);
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        assertInstanceOf(ScreenState.Content.class, state);
        ReportData rd = (ReportData) ((ScreenState.Content) state).getData();
        assertEquals(ReportPeriod.DAILY, rd.getPeriod());
    }
}

File: src/test/java/com/timeback/ui/viewmodel/ReportViewModelTest.kt
package com.timeback.ui.viewmodel

import com.timeback.ui.domain.model.*
import com.timeback.ui.fake.FakeFeatureGateway
import com.timeback.ui.feature.report.ReportViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.*
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*

/**
 * UI-07 ReportViewModel 단위 테스트
 * 검증: loadReport → CONTENT, changePeriod 전환
 */
@OptIn(ExperimentalCoroutinesApi::class)
class ReportViewModelTest {

    private val testDispatcher = UnconfinedTestDispatcher()
    private lateinit var gateway: FakeFeatureGateway
    private lateinit var viewModel: ReportViewModel

    @BeforeEach
    fun setup() {
        Dispatchers.setMain(testDispatcher)
        gateway = FakeFeatureGateway()
        viewModel = ReportViewModel(gateway)
    }

    @AfterEach
    fun tearDown() {
        Dispatchers.resetMain()
    }

    @Test
    fun `initial load shows CONTENT with report data`() = runTest {
        advanceUntilIdle()

        val state = viewModel.state.value
        assertTrue(state is ScreenState.Content<*>)
        val data = (state as ScreenState.Content<*>).data as ReportData
        assertEquals(ReportPeriod.WEEKLY, data.period)
        assertTrue(data.baselineComparable)
    }

    @Test
    fun `changePeriod reloads report`() = runTest {
        advanceUntilIdle()

        viewModel.changePeriod(ReportPeriod.DAILY)
        advanceUntilIdle()

        val state = viewModel.state.value
        assertTrue(state is ScreenState.Content<*>)
    }
}
